pkgs/test_core/lib/src/runner/vm/test_compiler.dart

```diff
@@ -0,0 +1,273 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+import 'dart:math' as math;
+import 'dart:typed_data';
+
+// ignore: import_of_legacy_library_into_null_safe
+import 'package:package_config/package_config.dart';
+import 'package:path/path.dart' as path;
+import 'package:test_api/backend.dart'; // ignore: deprecated_member_use
+
+/// A request to the [TestCompiler] for recompilation.
+class _CompilationRequest {
+  _CompilationRequest(this.mainUri, this.result, this.metadata);
+
+  Uri mainUri;
+  Metadata metadata;
+  Completer<String> result;
+}
+
+class TestCompiler {
+  TestCompiler(this._dillCachePath)
+      : _outputDillDirectory =
+            Directory.systemTemp.createTempSync('dart_test.') {
+    _outputDill = File(path.join(_outputDillDirectory.path, 'output.dill'));
+    _compilerController.stream.listen(_onCompilationRequest, onDone: () {
+      _outputDillDirectory.deleteSync(recursive: true);
+    });
+  }
+
+  final String _dillCachePath;
+  final Directory _outputDillDirectory;
+  final _compilerController = StreamController<_CompilationRequest>();
+  final _compilationQueue = <_CompilationRequest>[];
+  final _stdoutHandler = _StdoutHandler();
+
+  File? _outputDill;
+  Process? _compiler;
+  PackageConfig? _packageConfig;
+
+  Future<String?> compile(Uri mainDart, Metadata metadata) {
+    final completer = Completer<String>();
+    if (_compilerController.isClosed) {
+      return Future.value(null);
+    }
+    _compilerController.add(_CompilationRequest(mainDart, completer, metadata));
+    return completer.future;
+  }
+
+  Future<void> _shutdown() async {
+    if (_compiler != null) {
+      _compiler!.kill();
+      _compiler = null;
+    }
+  }
+
+  Future<void> dispose() async {
+    await _compilerController.close();
+    await _shutdown();
+  }
+
+  Future<String> _languageVersionComment(Uri testUri) async {
+    var localPackageConfig = _packageConfig ??= await loadPackageConfig(File(
```

Comment:
The existing language version API didn't work for some reason, maybe because of how I was running test from source from another project? It just picked up the test package config and not the project package config